print benchmark information for every layer, especially for CONVOLUTION

lhCheung1991 opened the pull request:
## Aim
Generally, Deep-Learning engineers usually pay more attention to network training, but neglect the performance of network inferencing, especially on mobile-platform.

These codes aim to help the engineers profile the performance and then shoot the bottleneck of the performance of their network. According to the profile result, they can adjust the network structure to make a trade-off between network accuracy and network speed on ncnn.

## Usage
Compile the ncnn with *_-DNCNN_BENCHMARK=ON_(OFF for default)*, for example, on armv7a & Mac
```shell
##### android armv7
mkdir -p build-android-armv7
pushd build-android-armv7
cmake   -DCMAKE_TOOLCHAIN_FILE=../android.toolchain.cmake \
        -DANDROID_ABI="armeabi-v7a with NEON" \
        -DANDROID_NATIVE_API_LEVEL=android-9 \
        -DANDROID_NDK=/Users/air/Documents/GitHub/android-ndk-r15c \
        -DANDROID_FORCE_ARM_BUILD=OFF \
        -DANDROID_STL_FORCE_FEATURES=OFF \
        -DNCNN_OPENMP=OFF \
        -DANDROID_TOOLCHAIN_NAME=arm-linux-androideabi-4.9 \
        -DNCNN_BENCHMARK=ON \
        ..
make -j 8
make install
popd
adb push build-android-armv7/bin/squeezenet /data/local/tmp

##### MacOS
mkdir -p build-mac
pushd build-mac
cmake   -DNCNN_OPENMP=OFF \
        -DNCNN_OPENCV=ON \
        -DNCNN_BENCHMARK=ON \
        ..
make -j 8
make install
popd 
```

## Result
A benchmark example of squeezenet.
Platform: Google NEXUS P5
Input: 3 x 224 x 224
```shell
Convolution              conv1                       15.05ms    |    feature_map:  224 x 224     inch:    3    outch:   64     kernel: 3 x 3     stride: 2 x 2
ReLU                     relu_conv1                   1.80ms    |    feature_map:  111 x 111     inch:   64    outch:   64
Pooling                  pool1                        3.13ms    |    feature_map:  111 x 111     inch:   64    outch:   64
Convolution              fire2/squeeze1x1             1.39ms    |    feature_map:   55 x 55      inch:   64    outch:   16     kernel: 1 x 1     stride: 1 x 1
ReLU                     fire2/relu_squeeze1x1        0.05ms    |    feature_map:   55 x 55      inch:   16    outch:   16
Split                    splitncnn_0                  0.00ms    |
Convolution              fire2/expand1x1              1.94ms    |    feature_map:   55 x 55      inch:   16    outch:   64     kernel: 1 x 1     stride: 1 x 1
ReLU                     fire2/relu_expand1x1         0.18ms    |    feature_map:   55 x 55      inch:   64    outch:   64
Convolution              fire2/expand3x3             13.30ms    |    feature_map:   55 x 55      inch:   16    outch:   64     kernel: 3 x 3     stride: 1 x 1
ReLU                     fire2/relu_expand3x3         0.28ms    |    feature_map:   55 x 55      inch:   64    outch:   64
Concat                   fire2/concat                 2.31ms    |
Convolution              fire3/squeeze1x1             2.72ms    |    feature_map:   55 x 55      inch:  128    outch:   16     kernel: 1 x 1     stride: 1 x 1
ReLU                     fire3/relu_squeeze1x1        0.06ms    |    feature_map:   55 x 55      inch:   16    outch:   16
Split                    splitncnn_1                  0.00ms    |
Convolution              fire3/expand1x1              1.94ms    |    feature_map:   55 x 55      inch:   16    outch:   64     kernel: 1 x 1     stride: 1 x 1
ReLU                     fire3/relu_expand1x1         0.29ms    |    feature_map:   55 x 55      inch:   64    outch:   64
Convolution              fire3/expand3x3             12.87ms    |    feature_map:   55 x 55      inch:   16    outch:   64     kernel: 3 x 3     stride: 1 x 1
ReLU                     fire3/relu_expand3x3         0.30ms    |    feature_map:   55 x 55      inch:   64    outch:   64
Concat                   fire3/concat                 2.24ms    |
Pooling                  pool3                        1.64ms    |    feature_map:   55 x 55      inch:  128    outch:  128
Convolution              fire4/squeeze1x1             1.38ms    |    feature_map:   27 x 27      inch:  128    outch:   32     kernel: 1 x 1     stride: 1 x 1
ReLU                     fire4/relu_squeeze1x1        0.02ms    |    feature_map:   27 x 27      inch:   32    outch:   32
Split                    splitncnn_2                  0.00ms    |
Convolution              fire4/expand1x1              1.60ms    |    feature_map:   27 x 27      inch:   32    outch:  128     kernel: 1 x 1     stride: 1 x 1
ReLU                     fire4/relu_expand1x1         0.07ms    |    feature_map:   27 x 27      inch:  128    outch:  128
Convolution              fire4/expand3x3              9.40ms    |    feature_map:   27 x 27      inch:   32    outch:  128     kernel: 3 x 3     stride: 1 x 1
ReLU                     fire4/relu_expand3x3         0.11ms    |    feature_map:   27 x 27      inch:  128    outch:  128
Concat                   fire4/concat                 1.09ms    |
Convolution              fire5/squeeze1x1             2.65ms    |    feature_map:   27 x 27      inch:  256    outch:   32     kernel: 1 x 1     stride: 1 x 1
ReLU                     fire5/relu_squeeze1x1        0.02ms    |    feature_map:   27 x 27      inch:   32    outch:   32
Split                    splitncnn_3                  0.00ms    |
Convolution              fire5/expand1x1              1.63ms    |    feature_map:   27 x 27      inch:   32    outch:  128     kernel: 1 x 1     stride: 1 x 1
ReLU                     fire5/relu_expand1x1         0.07ms    |    feature_map:   27 x 27      inch:  128    outch:  128
Convolution              fire5/expand3x3              9.39ms    |    feature_map:   27 x 27      inch:   32    outch:  128     kernel: 3 x 3     stride: 1 x 1
ReLU                     fire5/relu_expand3x3         0.11ms    |    feature_map:   27 x 27      inch:  128    outch:  128
Concat                   fire5/concat                 1.16ms    |
Pooling                  pool5                        0.68ms    |    feature_map:   27 x 27      inch:  256    outch:  256
Convolution              fire6/squeeze1x1             0.70ms    |    feature_map:   13 x 13      inch:  256    outch:   48     kernel: 1 x 1     stride: 1 x 1
ReLU                     fire6/relu_squeeze1x1        0.01ms    |    feature_map:   13 x 13      inch:   48    outch:   48
Split                    splitncnn_4                  0.00ms    |
Convolution              fire6/expand1x1              0.66ms    |    feature_map:   13 x 13      inch:   48    outch:  192     kernel: 1 x 1     stride: 1 x 1
ReLU                     fire6/relu_expand1x1         0.02ms    |    feature_map:   13 x 13      inch:  192    outch:  192
Convolution              fire6/expand3x3              6.39ms    |    feature_map:   13 x 13      inch:   48    outch:  192     kernel: 3 x 3     stride: 1 x 1
ReLU                     fire6/relu_expand3x3         0.02ms    |    feature_map:   13 x 13      inch:  192    outch:  192
Concat                   fire6/concat                 0.39ms    |
Convolution              fire7/squeeze1x1             0.96ms    |    feature_map:   13 x 13      inch:  384    outch:   48     kernel: 1 x 1     stride: 1 x 1
ReLU                     fire7/relu_squeeze1x1        0.01ms    |    feature_map:   13 x 13      inch:   48    outch:   48
Split                    splitncnn_5                  0.00ms    |
Convolution              fire7/expand1x1              0.69ms    |    feature_map:   13 x 13      inch:   48    outch:  192     kernel: 1 x 1     stride: 1 x 1
ReLU                     fire7/relu_expand1x1         0.03ms    |    feature_map:   13 x 13      inch:  192    outch:  192
Convolution              fire7/expand3x3              6.27ms    |    feature_map:   13 x 13      inch:   48    outch:  192     kernel: 3 x 3     stride: 1 x 1
ReLU                     fire7/relu_expand3x3         0.02ms    |    feature_map:   13 x 13      inch:  192    outch:  192
Concat                   fire7/concat                 0.38ms    |
Convolution              fire8/squeeze1x1             1.24ms    |    feature_map:   13 x 13      inch:  384    outch:   64     kernel: 1 x 1     stride: 1 x 1
ReLU                     fire8/relu_squeeze1x1        0.01ms    |    feature_map:   13 x 13      inch:   64    outch:   64
Split                    splitncnn_6                  0.00ms    |
Convolution              fire8/expand1x1              1.12ms    |    feature_map:   13 x 13      inch:   64    outch:  256     kernel: 1 x 1     stride: 1 x 1
ReLU                     fire8/relu_expand1x1         0.03ms    |    feature_map:   13 x 13      inch:  256    outch:  256
Convolution              fire8/expand3x3             10.17ms    |    feature_map:   13 x 13      inch:   64    outch:  256     kernel: 3 x 3     stride: 1 x 1
ReLU                     fire8/relu_expand3x3         0.03ms    |    feature_map:   13 x 13      inch:  256    outch:  256
Concat                   fire8/concat                 0.51ms    |
Convolution              fire9/squeeze1x1             1.69ms    |    feature_map:   13 x 13      inch:  512    outch:   64     kernel: 1 x 1     stride: 1 x 1
ReLU                     fire9/relu_squeeze1x1        0.01ms    |    feature_map:   13 x 13      inch:   64    outch:   64
Split                    splitncnn_7                  0.00ms    |
Convolution              fire9/expand1x1              1.07ms    |    feature_map:   13 x 13      inch:   64    outch:  256     kernel: 1 x 1     stride: 1 x 1
ReLU                     fire9/relu_expand1x1         0.03ms    |    feature_map:   13 x 13      inch:  256    outch:  256
Convolution              fire9/expand3x3             10.07ms    |    feature_map:   13 x 13      inch:   64    outch:  256     kernel: 3 x 3     stride: 1 x 1
ReLU                     fire9/relu_expand3x3         0.03ms    |    feature_map:   13 x 13      inch:  256    outch:  256
Concat                   fire9/concat                 0.54ms    |
Dropout                  drop9                        0.00ms    |    feature_map:   13 x 13      inch:  512    outch:  512
Convolution              conv10                      35.71ms    |    feature_map:   13 x 13      inch:  512    outch: 1000     kernel: 1 x 1     stride: 1 x 1
ReLU                     relu_conv10                  0.32ms    |    feature_map:   15 x 15      inch: 1000    outch: 1000
Pooling                  pool10                       0.43ms    |    feature_map:   15 x 15      inch: 1000    outch: 1000
Softmax                  prob                         0.12ms    |    feature_map:    1 x 1       inch: 1000    outch: 1000
```

nihui commented:
hello
It is not suitable to use clock function to get the timestamp that is unstable in multithreading programs I think.

reference
https://stackoverflow.com/questions/2962785/c-using-clock-to-measure-time-in-multi-threaded-programs

nihui commented:
need to implement a platform-independent timestamp function for this purpose ...
gettimeofday for linux and something for windows ...

lhCheung1991 commented:
Thanks for your precious opinions, I found that *_chrono::steady_clock_* from C++11 is quite suitable for this situation, which is a monotonic and nanosecond-resolution cross a various OS. 
[]()https://solarianprogrammer.com/2012/10/14/cpp-11-timing-code-performance/
[]()http://en.cppreference.com/w/cpp/chrono/steady_clock

nihui commented:
ncnn codebase does not depend on c++11 features  :P

lhCheung1991 commented:
Got it. Let me find a better solution for it.

lhCheung1991 commented:
Hi, I have updated the commit.
## Update for cross-platform
I use `gettimeofday()` from <sys/time.h> for Unix-like system and a fake `gettimeofday()` defined as follow for Windows:
```c++
typedef struct timeval 
{
    long tv_sec;
    long tv_usec;
} timeval;

int gettimeofday(struct timeval * tp, struct timezone * tzp)
{
    // Note: some broken versions only have 8 trailing zero's, the correct epoch has 9 trailing zero's
    // This magic number is the number of 100 nanosecond intervals since January 1, 1601 (UTC)
    // until 00:00:00 January 1, 1970 
    static const uint64_t EPOCH = ((uint64_t) 116444736000000000ULL);

    SYSTEMTIME  system_time;
    FILETIME    file_time;
    uint64_t    time;

    GetSystemTime( &system_time );
    SystemTimeToFileTime( &system_time, &file_time );
    time =  ((uint64_t)file_time.dwLowDateTime )      ;
    time += ((uint64_t)file_time.dwHighDateTime) << 32;

    tp->tv_sec  = (long) ((time - EPOCH) / 10000000L);
    tp->tv_usec = (long) (system_time.wMilliseconds * 1000);
    return 0;
}
```
## Updated Result
A benchmark example of squeezenet.
Platform: Google NEXUS P5
Input: 3 x 224 x 224
```shell
Convolution              conv1                       12.54ms    |    feature_map:  224 x 224     inch:    3    outch:   64     kernel: 3 x 3     stride: 2 x 2
ReLU                     relu_conv1                   2.19ms    |    feature_map:  111 x 111     inch:   64    outch:   64
Pooling                  pool1                        3.58ms    |    feature_map:  111 x 111     inch:   64    outch:   64
Convolution              fire2/squeeze1x1             1.61ms    |    feature_map:   55 x 55      inch:   64    outch:   16     kernel: 1 x 1     stride: 1 x 1
ReLU                     fire2/relu_squeeze1x1        0.03ms    |    feature_map:   55 x 55      inch:   16    outch:   16
Split                    splitncnn_0                  0.00ms    |
Convolution              fire2/expand1x1              2.13ms    |    feature_map:   55 x 55      inch:   16    outch:   64     kernel: 1 x 1     stride: 1 x 1
ReLU                     fire2/relu_expand1x1         0.24ms    |    feature_map:   55 x 55      inch:   64    outch:   64
Convolution              fire2/expand3x3             13.77ms    |    feature_map:   55 x 55      inch:   16    outch:   64     kernel: 3 x 3     stride: 1 x 1
ReLU                     fire2/relu_expand3x3         0.24ms    |    feature_map:   55 x 55      inch:   64    outch:   64
Concat                   fire2/concat                 2.50ms    |
Convolution              fire3/squeeze1x1             2.88ms    |    feature_map:   55 x 55      inch:  128    outch:   16     kernel: 1 x 1     stride: 1 x 1
ReLU                     fire3/relu_squeeze1x1        0.05ms    |    feature_map:   55 x 55      inch:   16    outch:   16
Split                    splitncnn_1                  0.00ms    |
Convolution              fire3/expand1x1              2.13ms    |    feature_map:   55 x 55      inch:   16    outch:   64     kernel: 1 x 1     stride: 1 x 1
ReLU                     fire3/relu_expand1x1         0.25ms    |    feature_map:   55 x 55      inch:   64    outch:   64
Convolution              fire3/expand3x3             13.90ms    |    feature_map:   55 x 55      inch:   16    outch:   64     kernel: 3 x 3     stride: 1 x 1
ReLU                     fire3/relu_expand3x3         0.27ms    |    feature_map:   55 x 55      inch:   64    outch:   64
Concat                   fire3/concat                 2.51ms    |
Pooling                  pool3                        1.84ms    |    feature_map:   55 x 55      inch:  128    outch:  128
Convolution              fire4/squeeze1x1             1.50ms    |    feature_map:   27 x 27      inch:  128    outch:   32     kernel: 1 x 1     stride: 1 x 1
ReLU                     fire4/relu_squeeze1x1        0.01ms    |    feature_map:   27 x 27      inch:   32    outch:   32
Split                    splitncnn_2                  0.00ms    |
Convolution              fire4/expand1x1              1.76ms    |    feature_map:   27 x 27      inch:   32    outch:  128     kernel: 1 x 1     stride: 1 x 1
ReLU                     fire4/relu_expand1x1         0.05ms    |    feature_map:   27 x 27      inch:  128    outch:  128
Convolution              fire4/expand3x3              9.72ms    |    feature_map:   27 x 27      inch:   32    outch:  128     kernel: 3 x 3     stride: 1 x 1
ReLU                     fire4/relu_expand3x3         0.07ms    |    feature_map:   27 x 27      inch:  128    outch:  128
Concat                   fire4/concat                 1.18ms    |
Convolution              fire5/squeeze1x1             2.87ms    |    feature_map:   27 x 27      inch:  256    outch:   32     kernel: 1 x 1     stride: 1 x 1
ReLU                     fire5/relu_squeeze1x1        0.01ms    |    feature_map:   27 x 27      inch:   32    outch:   32
Split                    splitncnn_3                  0.00ms    |
Convolution              fire5/expand1x1              1.79ms    |    feature_map:   27 x 27      inch:   32    outch:  128     kernel: 1 x 1     stride: 1 x 1
ReLU                     fire5/relu_expand1x1         0.06ms    |    feature_map:   27 x 27      inch:  128    outch:  128
Convolution              fire5/expand3x3             10.18ms    |    feature_map:   27 x 27      inch:   32    outch:  128     kernel: 3 x 3     stride: 1 x 1
ReLU                     fire5/relu_expand3x3         0.06ms    |    feature_map:   27 x 27      inch:  128    outch:  128
Concat                   fire5/concat                 1.31ms    |
Pooling                  pool5                        0.74ms    |    feature_map:   27 x 27      inch:  256    outch:  256
Convolution              fire6/squeeze1x1             0.73ms    |    feature_map:   13 x 13      inch:  256    outch:   48     kernel: 1 x 1     stride: 1 x 1
ReLU                     fire6/relu_squeeze1x1        0.01ms    |    feature_map:   13 x 13      inch:   48    outch:   48
Split                    splitncnn_4                  0.00ms    |
Convolution              fire6/expand1x1              0.78ms    |    feature_map:   13 x 13      inch:   48    outch:  192     kernel: 1 x 1     stride: 1 x 1
ReLU                     fire6/relu_expand1x1         0.03ms    |    feature_map:   13 x 13      inch:  192    outch:  192
Convolution              fire6/expand3x3              7.75ms    |    feature_map:   13 x 13      inch:   48    outch:  192     kernel: 3 x 3     stride: 1 x 1
ReLU                     fire6/relu_expand3x3         0.02ms    |    feature_map:   13 x 13      inch:  192    outch:  192
Concat                   fire6/concat                 0.34ms    |
Convolution              fire7/squeeze1x1             0.91ms    |    feature_map:   13 x 13      inch:  384    outch:   48     kernel: 1 x 1     stride: 1 x 1
ReLU                     fire7/relu_squeeze1x1        0.01ms    |    feature_map:   13 x 13      inch:   48    outch:   48
Split                    splitncnn_5                  0.00ms    |
Convolution              fire7/expand1x1              0.66ms    |    feature_map:   13 x 13      inch:   48    outch:  192     kernel: 1 x 1     stride: 1 x 1
ReLU                     fire7/relu_expand1x1         0.02ms    |    feature_map:   13 x 13      inch:  192    outch:  192
Convolution              fire7/expand3x3              6.20ms    |    feature_map:   13 x 13      inch:   48    outch:  192     kernel: 3 x 3     stride: 1 x 1
ReLU                     fire7/relu_expand3x3         0.02ms    |    feature_map:   13 x 13      inch:  192    outch:  192
Concat                   fire7/concat                 0.34ms    |
Convolution              fire8/squeeze1x1             1.20ms    |    feature_map:   13 x 13      inch:  384    outch:   64     kernel: 1 x 1     stride: 1 x 1
ReLU                     fire8/relu_squeeze1x1        0.01ms    |    feature_map:   13 x 13      inch:   64    outch:   64
Split                    splitncnn_6                  0.00ms    |
Convolution              fire8/expand1x1              1.04ms    |    feature_map:   13 x 13      inch:   64    outch:  256     kernel: 1 x 1     stride: 1 x 1
ReLU                     fire8/relu_expand1x1         0.03ms    |    feature_map:   13 x 13      inch:  256    outch:  256
Convolution              fire8/expand3x3             10.10ms    |    feature_map:   13 x 13      inch:   64    outch:  256     kernel: 3 x 3     stride: 1 x 1
ReLU                     fire8/relu_expand3x3         0.03ms    |    feature_map:   13 x 13      inch:  256    outch:  256
Concat                   fire8/concat                 0.48ms    |
Convolution              fire9/squeeze1x1             1.61ms    |    feature_map:   13 x 13      inch:  512    outch:   64     kernel: 1 x 1     stride: 1 x 1
ReLU                     fire9/relu_squeeze1x1        0.01ms    |    feature_map:   13 x 13      inch:   64    outch:   64
Split                    splitncnn_7                  0.00ms    |
Convolution              fire9/expand1x1              1.04ms    |    feature_map:   13 x 13      inch:   64    outch:  256     kernel: 1 x 1     stride: 1 x 1
ReLU                     fire9/relu_expand1x1         0.03ms    |    feature_map:   13 x 13      inch:  256    outch:  256
Convolution              fire9/expand3x3             10.05ms    |    feature_map:   13 x 13      inch:   64    outch:  256     kernel: 3 x 3     stride: 1 x 1
ReLU                     fire9/relu_expand3x3         0.03ms    |    feature_map:   13 x 13      inch:  256    outch:  256
Concat                   fire9/concat                 0.48ms    |
Dropout                  drop9                        0.00ms    |    feature_map:   13 x 13      inch:  512    outch:  512
Convolution              conv10                      33.45ms    |    feature_map:   13 x 13      inch:  512    outch: 1000     kernel: 1 x 1     stride: 1 x 1
ReLU                     relu_conv10                  0.34ms    |    feature_map:   15 x 15      inch: 1000    outch: 1000
Pooling                  pool10                       0.36ms    |    feature_map:   15 x 15      inch: 1000    outch: 1000
Softmax                  prob                         0.11ms    |    feature_map:    1 x 1       inch: 1000    outch: 1000
```

nihui commented on src/benchmark.h:
```diff
@@ -0,0 +1,63 @@
+    long tv_usec;
+} timeval;
+
+int gettimeofday(struct timeval * tp, struct timezone * tzp)
```
function implementation should be moved to cpp file.
otherwise, linker will complain about multiple definitions if several cpp files include this header.

lhCheung1991 replied:
Function implementaion of `gettimeofday()` for Windows has been moved to benchmark.cpp

lhCheung1991 commented:
I see. You made an alias for the timestamp functions and manage them within the `namespace ncnn` to avoid the potential linking conflict due to the wide usage of signature `int gettimeofday(struct timeval *restrict tp, void *restrict tzp);`. Quite a considerate design. 

SALUTE TO __Nihui the Great__ !

AngryPowman commented:
It's very helpful to me

Deeperfinder commented:
How to printf the benchmark of my neural networks?
